File: mobile/airmouse-mobile/src/engine/filters.ts
// One Euro Filter - Port from Python AirMouse
// Reference: https://hal.inria.fr/hal-00670496/document

class LowPass {
  private y: number | null = null;

  apply(x: number, alpha: number): number {
    this.y = this.y === null ? x : alpha * x + (1.0 - alpha) * this.y;
    return this.y;
  }

  reset(): void {
    this.y = null;
  }
}

export class OneEuroFilter {
  private minCutoff: number;
  private beta: number;
  private dCutoff: number;
  private velocity: number = 0.0;
  private xLpf = new LowPass();
  private dxLpf = new LowPass();
  private xPrev: number | null = null;
  private tPrev: number | null = null;

  constructor(minCutoff: number = 1.4, beta: number = 0.028, dCutoff: number = 1.0) {
    this.minCutoff = minCutoff;
    this.beta = beta;
    this.dCutoff = dCutoff;
  }

  reset(): void {
    this.xLpf.reset();
    this.dxLpf.reset();
    this.xPrev = null;
    this.tPrev = null;
    this.velocity = 0.0;
  }

  private static alpha(cutoff: number, dt: number): number {
    const tau = 1.0 / (2.0 * Math.PI * cutoff);
    return 1.0 / (1.0 + tau / dt);
  }

  filter(x: number, t?: number): number {
    const now = t ?? performance.now() / 1000;
    let dt: number;
    if (this.tPrev === null) {
      dt = 1.0 / 30.0;
    } else {
      dt = Math.max(now - this.tPrev, 1e-6);
    }
    this.tPrev = now;

    const dx = this.xPrev === null ? 0.0 : (x - this.xPrev) / dt;
    this.xPrev = x;
    const edx = this.dxLpf.apply(dx, OneEuroFilter.alpha(this.dCutoff, dt));
    this.velocity = edx;
    const cutoff = this.minCutoff + this.beta * Math.abs(edx);
    return this.xLpf.apply(x, OneEuroFilter.alpha(cutoff, dt));
  }

  getVelocity(): number {
    return this.velocity;
  }
}

export class FilterPair2D {
  private fx: OneEuroFilter;
  private fy: OneEuroFilter;
  vx: number = 0.0;
  vy: number = 0.0;

  constructor(minCutoff: number = 1.4, beta: number = 0.028) {
    this.fx = new OneEuroFilter(minCutoff, beta);
    this.fy = new OneEuroFilter(minCutoff, beta);
  }

  get velocity(): number {
    return Math.hypot(this.vx, this.vy);
  }

  setParams(minCutoff: number, beta: number): void {
    this.fx['minCutoff'] = minCutoff;
    this.fx['beta'] = beta;
    this.fy['minCutoff'] = minCutoff;
    this.fy['beta'] = beta;
  }

  reset(): void {
    this.fx.reset();
    this.fy.reset();
    this.vx = 0.0;
    this.vy = 0.0;
  }

  filter(x: number, y: number): [number, number] {
    const rx = this.fx.filter(x);
    const ry = this.fy.filter(y);
    this.vx = this.fx.getVelocity();
    this.vy = this.fy.getVelocity();
    return [rx, ry];
  }
}

export class AccelCurve {
  minGain: number;
  maxGain: number;
  refSpeed: number;
  expo: number;

  constructor(minGain: number = 1.2, maxGain: number = 3.0, refSpeed: number = 1400.0, expo: number = 1.7) {
    this.minGain = minGain;
    this.maxGain = maxGain;
    this.refSpeed = Math.max(refSpeed, 1e-6);
    this.expo = expo;
  }

  apply(vx: number, vy: number): number {
    const t = Math.min(Math.hypot(vx, vy) / this.refSpeed, 1.0);
    let s: number;
    if (this.expo > 0.0) {
      s = Math.pow(t, this.expo);
    } else {
      s = t * t * (3.0 - 2.0 * t);
    }
    return this.minGain + (this.maxGain - this.minGain) * s;
  }
}

File: core/filters.py
import math
import time


class _LowPass:
    def __init__(self):
        self.y = None

    def apply(self, x, alpha):
        self.y = x if self.y is None else alpha * x + (1.0 - alpha) * self.y
        return self.y


class OneEuroFilter:
    def __init__(self, min_cutoff=1.4, beta=0.028, d_cutoff=1.0):
        self.min_cutoff = min_cutoff
        self.beta = beta
        self.d_cutoff = d_cutoff
        self.velocity = 0.0
        self._x_lpf = _LowPass()
        self._dx_lpf = _LowPass()
        self._x_prev = None
        self._t_prev = None

    def reset(self):
        self._x_lpf.y = None
        self._dx_lpf.y = None
        self._x_prev = None
        self._t_prev = None
        self.velocity = 0.0

    @staticmethod
    def _alpha(cutoff, dt):
        tau = 1.0 / (2.0 * math.pi * cutoff)
        return 1.0 / (1.0 + tau / dt)

    def filter(self, x, t=None):
        now = time.perf_counter() if t is None else t
        if self._t_prev is None:
            dt = 1.0 / 30.0
        else:
            dt = max(now - self._t_prev, 1e-6)
        self._t_prev = now

        dx = 0.0 if self._x_prev is None else (x - self._x_prev) / dt
        self._x_prev = x
        edx = self._dx_lpf.apply(dx, self._alpha(self.d_cutoff, dt))
        self.velocity = edx
        cutoff = self.min_cutoff + self.beta * abs(edx)
        return self._x_lpf.apply(x, self._alpha(cutoff, dt))


class FilterPair2D:
    def __init__(self, min_cutoff=1.4, beta=0.028):
        self.fx = OneEuroFilter(min_cutoff=min_cutoff, beta=beta)
        self.fy = OneEuroFilter(min_cutoff=min_cutoff, beta=beta)
        self.vx = 0.0
        self.vy = 0.0

    @property
    def velocity(self):
        return math.hypot(self.vx, self.vy)

    def set_params(self, min_cutoff, beta):
        for f in (self.fx, self.fy):
            f.min_cutoff = min_cutoff
            f.beta = beta

    def reset(self):
        self.fx.reset()
        self.fy.reset()
        self.vx = 0.0
        self.vy = 0.0

    def filter(self, x, y):
        rx = self.fx.filter(x)
        ry = self.fy.filter(y)
        self.vx = self.fx.velocity
        self.vy = self.fy.velocity
        return rx, ry


class AccelCurve:
    """Curva de aceleracao tipo rato gaming.

    expo <= 0 mantem o smoothstep classico; expo > 0 usa curva de potencia
    (t**expo), que e mais precisa devagar e mais agressiva a varrer.
    """

    def __init__(self, min_gain=1.2, max_gain=3.0, ref_speed=1400.0, expo=1.7):
        self.min_gain = float(min_gain)
        self.max_gain = float(max_gain)
        self.ref_speed = max(float(ref_speed), 1e-6)
        self.expo = float(expo)

    def apply(self, vx, vy):
        t = min(math.hypot(vx, vy) / self.ref_speed, 1.0)
        if self.expo > 0.0:
            s = t ** self.expo
        else:
            s = t * t * (3.0 - 2.0 * t)
        return self.min_gain + (self.max_gain - self.min_gain) * s
